### PR TITLE
Polymorphic stackoverflow prevention

### LINKED ISSUE #591: component cycle stub generation StackOverflowError
**Description**
Generating stubs and tests for OpenApi spec containing component cycles causes StackOverflowError.

**Steps to reproduce**
1. See unit tests in pull request #608 

**Expected behavior**
Stub generation and contract testing without error

**Screenshots**
n/a

**System Information:**
 - OS & version: Ubuntu
 - Browser & version: n/a
 - Specmatic version: 0.59.0
 - JDK version: 11

**Additional context**

### PR DESCRIPTION
**Pull request by westse**
**What**: More complete stackoverflow prevention than has been implemented so far

**Why**: Fixes #591. I kept running into new scenarios that produced StackOverflowError, so I felt it was time for a more complete refactor after taking the time to fully understand the OpenApiSpecification class logic. Additionally, cycle prevention logic is needed in stub and test generation logic and has been included here (used to be in #592 which is now closed in favor of this PR)

**How**:
 Pattern discovery (OpenApiSpecification)
- Return previously calculated component patterns rather than calculating them again.
- Standardize how the component pattern cache is populated by factoring into a dedicated method.
- Simplify toSchemaProperties() logic to remove logic that is redundant to existing toSpecmaticPattern() logic.
- Ensure typeStack is populated with proper component pattern names in all the correct recursive calls.

Stub generation (`Pattern.generate()`) and test generation (`Pattern.newBasedOn()`)
- For stub and test generation, `Resolver.withCyclePrevention()` is introduced as a wrapper for any call that resolves deferred patterns to actual patterns.  This is an approach generally applicable for bot test and stub generation.
- Ensures Resolver remains immutable via `Resolver.withCyclePrevention()` which produces a new Resolver with a cyclePreventionStack with an addition entry for the current pattern.
- `Resolver.withCyclePrevention()` has a simple overload which will throw ContractException(cycle=true) if a cycle is found, but there is also a second overload that allows returning null to give the caller an opportunity to gracefully handle the detected cycle. This is done for nullable components (see AnyPattern) and optional properties (see JSONObjectPattern).

**Checklist**:

- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation
- [X] Tests
- [ ] Sonar Quality Gate

**Issue ID**:
Closes: #591

### CODE REVIEW

File: core/src/test/kotlin/in/specmatic/conversions/OpenApiKtTest.kt [hunk 1]
```diff
@@ -1022,18 +1131,125 @@ Background:
...
+                    .addHeader("Content-Type", "application/json")
+                    .get().build()
+            val call = OkHttpClient().newCall(request)
+            call.execute()
+        }
+
+        assertThat(resp.isSuccessful).isFalse
+        assertThat(resp.code()).isEqualTo(400)
+        val body = resp.body()?.string()
+        assertThat(body).contains("Invalid pattern cycle")
```
Comment (harikrishnan83) [inline core/src/test/kotlin/in/specmatic/conversions/OpenApiKtTest.kt:1160]: This is important, I am wondering if there is a way in the parser library itself to detect the unsatisfiable circular dependency (Mandatory non-nullable reference that leads to an unbreakable chain). However at this point this fix is a good solution. cc @joelrosario

Comment (harikrishnan83) [reply core/src/test/kotlin/in/specmatic/conversions/OpenApiKtTest.kt:1160]: Good that this also logs stack that is responsible for the circular reference

File: core/src/test/kotlin/in/specmatic/conversions/OpenApiKtTest.kt [hunk 2]
```diff
@@ -1022,18 +1131,125 @@ Background:
...
+        val feature = parseGherkinStringToFeature(
+            """
+Feature: Hello world
+
+Background:
+  Given openapi openapi/circular-reference-optional-non-nullable.yaml
+        """.trimIndent(), sourceSpecPath
+        )
+
+        val result = testBackwardCompatibility(feature, feature)
```
Comment (harikrishnan83) [inline core/src/test/kotlin/in/specmatic/conversions/OpenApiKtTest.kt:1175]: Not sure I fully understand this. Seems to be testing both backward compatibility (to make sure feature is compatible with itself) and also testing Stubbing.

File: core/src/test/kotlin/in/specmatic/conversions/OpenApiKtTest.kt [hunk 3]
```diff
@@ -1022,18 +1131,125 @@ Background:
...
+    @Test
+    @RepeatedTest(10) // Try to exercise all outcomes of AnyPattern.generate() which randomly selects from its options
+    fun `should validate and generate with polymorphic cyclic reference in open api`() {
         val feature = parseGherkinStringToFeature(
             """
 Feature: Hello world
 
 Background:
-  Given openapi openapi/circular-reference.yaml
+  Given openapi openapi/circular-reference-polymorphic.yaml
```
Comment (harikrishnan83) [inline core/src/test/kotlin/in/specmatic/conversions/OpenApiKtTest.kt:1233]: This is a really well designed test data. cc @joelrosario

**Comment by joelrosario** [comment]
Hi @westse , thanks for this, we really appreciate the effort you've put in to it. It's been on our bucket list for some time.

We have reviewed it and mostly understood it. We feel the need to document our understanding a little better with the help of a few more tests, which we shall write and push to your PR.

Once done, we'll merge it.

**Review by harikrishnan83 — APPROVED** [review]

**Pull request merged**